<a href="https://colab.research.google.com/github/gyanellaagreda-bd/AnaliticaEmpresarialIntegrada-26.2/blob/main/LAB_D6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ANALÍTICA EMPRESARIAL INTEGRADA
## LABORATORIO DIRIGIDO N.° 06

### DISEÑO Y ANÁLISIS DE UN EXPERIMENTO A/B

**Semana 6 · TECSUP 2026-II**  
**Docente:** Pilar Rocío Sayán Mejía  
**Carrera:** Big Data y Ciencia de Datos  
**Plataforma de trabajo:** Google Colab

### Pregunta orientadora

**¿Qué evidencia estadística y práctica necesitamos antes de implementar una variante B?**

> Este notebook acompaña el trabajo desarrollado durante la clase. No es un informe ni un proyecto para exponer. Las celdas alternan explicación breve, ejecución guiada, interpretación y retos de cierre.


# I. Elemento de la capacidad terminal de la semana

Diseña y analiza un experimento A/B identificando el factor, sus niveles, el control, el tratamiento, la unidad experimental y la variable respuesta. Formula las hipótesis estadísticas, establece la métrica primaria, los guardrails, el nivel de significación, la potencia y el efecto mínimo detectable; finalmente, integra la evidencia estadística con la relevancia práctica para sustentar una decisión.


# II. Seguridad y condiciones de trabajo

- Utilice únicamente Google Colab y las fuentes indicadas por la docente.
- No modifique ni fabrique resultados para forzar una conclusión.
- Ejecute las celdas en el orden propuesto y documente cualquier error o dato ausente.
- No comparta claves, credenciales ni información personal en el notebook.
- Mantenga la estación de trabajo ordenada y no instale software no autorizado.


# III. Fundamento teórico
Un A/B test compara un control A con una variante B mediante asignación aleatoria. La unidad experimental de este caso es el usuario, que conserva su variante durante toda la observación. La aleatorización favorece la comparabilidad en expectativa; no garantiza grupos idénticos.

La respuesta es binaria: compra dentro de los siete días posteriores a la asignación. Todos los usuarios completan esa ventana antes del análisis. La métrica primaria es la proporción de usuarios que compran. Se estudia la diferencia de dos proporciones, no ANOVA.

El MDE es el efecto utilizado para planificar la potencia. El umbral empresarial es la mejora que justificaría interés práctico. Aquí son distintos: MDE de 1,5 pp y umbral empresarial de 1,0 pp. La muestra se dimensiona para detectar el MDE frente a cero, no para garantizar evidencia de superar el umbral empresarial.

Se usa una aproximación normal de Wald para la diferencia. El contraste unilateral al 5 % y el límite inferior unilateral del 95 % utilizan el mismo error estándar. También se muestra un IC bilateral del 95 % como descripción de incertidumbre; no se exige que este último excluya cero para reproducir el contraste unilateral.

La recomendación académica requiere evidencia de superar el umbral empresarial y de mantener el guardrail dentro del margen predefinido. Son criterios más exigentes que obtener p < 0,05 frente a cero. Un resultado no concluyente no demuestra igualdad.


# IV. Normas y buenas prácticas empleadas

- Prerregistro lógico: definir hipótesis, métrica, MDE y regla de decisión antes de observar los resultados.
- Aleatorización: evitar que la asignación dependa de características de las unidades.
- Trazabilidad: conservar la fuente, las transformaciones y los parámetros utilizados.
- Reproducibilidad: ejecutar el análisis mediante código verificable.
- No realizar *peeking*: no detener o prolongar el experimento solo porque el resultado momentáneo conviene.
- No contaminar grupos: una unidad debe recibir únicamente la variante asignada.
- Distinguir asociación, evidencia experimental y decisión práctica.


# V. Recursos
- Computadora, Google Colab y notebook LAB-D6-AEI-PSAYAN-2026-02.ipynb.
- Librerías numpy, pandas, scipy, statsmodels y plotly.
- Caso ficticio NexaCommerce con datos simulados generados en el propio notebook. No requiere descargar un dataset externo.

## Naturaleza de los datos utilizados
Los registros son sintéticos y reproducibles mediante una semilla fija. Representan usuarios de un checkout; no proceden de una empresa real. Las probabilidades del simulador son supuestos académicos, no evidencia observada. No cambie la semilla para buscar un resultado favorable. En un estudio real, los datos procederían de registros del experimento y el plan se documentaría antes de recogerlos.


# VI. Metodología para el desarrollo de la sesión

El laboratorio se desarrolla mediante una secuencia guiada de cinco momentos:

1. **Comprender:** recuperar el vocabulario esencial.
2. **Representar:** traducir el caso a la estructura de un diseño experimental.
3. **Formular:** establecer las hipótesis y la regla de decisión antes de observar resultados.
4. **Ejecutar e interpretar:** correr el código por etapas y explicar cada salida.
5. **Decidir y transferir:** integrar evidencia estadística, umbral empresarial y guardrails; luego resolver retos breves.

No se solicita exposición, informe ni trabajo adicional. La evidencia de aprendizaje se construye durante la clase mediante respuestas breves en las pausas de análisis.


# VII. Procedimiento

## Actividad 1 — Activación conceptual

**Propósito:** reconocer los componentes que deben definirse antes de ejecutar un experimento.

| Concepto | Definición con sus propias palabras | Pregunta de apoyo |
|---|---|---|
| Experimento A/B | Es una prueba en la que se comparan dos versiones o condiciones, asignando las unidades de forma controlada para medir si un cambio produce un efecto diferente. | ¿Qué diferencia existe entre comparar dos grupos y experimentar? |
| Control | Es el grupo que mantiene la condición actual o de referencia y sirve como punto de comparación frente al tratamiento. | ¿Qué representa el punto de referencia? |
| Tratamiento | Es el grupo que recibe la nueva versión, cambio o intervención que se quiere evaluar. | ¿Qué cambio recibe este grupo? |
| Unidad experimental | Es la persona, cliente, usuario, transacción u otro elemento sobre el cual se aplica el tratamiento o control y se mide el resultado. | ¿Sobre quién o qué se aplica la variante? |
| Variable respuesta | Es el resultado que se observa y mide después de aplicar las condiciones del experimento a cada unidad experimental. | ¿Qué resultado se observa en cada unidad? |
| Métrica primaria | Es el indicador principal que se utilizará para determinar si el experimento logró el objetivo planteado. | ¿Cuál será el criterio principal de éxito? |
| Guardrail | Es una métrica de seguridad que permite verificar que la mejora de la métrica principal no provoque un deterioro importante en otro aspecto del negocio. | ¿Qué no debería deteriorarse? |
| MDE | Es el efecto mínimo que se considera relevante para el negocio y que el experimento debe tener capacidad de detectar. | ¿Qué efecto se utiliza para planificar la potencia? |
| Potencia | Es la capacidad estadística del experimento para detectar un efecto real del tamaño del MDE cuando dicho efecto existe. | ¿Qué capacidad tiene el diseño para detectar el MDE? |
| Peeking | Es revisar repetidamente los resultados antes de finalizar el experimento y tomar decisiones anticipadas, lo que puede aumentar el riesgo de obtener conclusiones incorrectas. | ¿Por qué no se debe decidir mirando resultados parciales? |

**Pausa de verificación:** explique a un compañero por qué el MDE no debe elegirse después de conocer la diferencia observada.


## Actividad 2 — Interpretación y representación metodológica
NexaCommerce, empresa ficticia, compara el checkout actual de cuatro pasos con uno simplificado de dos pasos. Cada usuario recibe una sola variante. Se mide compra e incidencia de pago dentro de siete días; incluso quienes no compran permanecen en el denominador.

### 2.1 Interpretación
Complete antes de ejecutar: Determinar si ____________________________________.

### 2.2 Representación del diseño
| Elemento | Definición en este experimento |
|---|---|
| Factor y niveles | Checkout: A actual; B simplificado |
| Unidad experimental | Usuario único con asignación persistente |
| Respuesta primaria | compra_7d: 1 si compró dentro de siete días; 0 si no |
| Métrica primaria | Usuarios que compraron / todos los usuarios asignados |
| Guardrail | incidencia_pago_7d: al menos una incidencia por usuario en siete días |
| Denominador del guardrail | Todos los usuarios asignados, compren o no |
| Margen de deterioro | B−A debe ser inferior a +0,5 pp en incidencias |
| Baseline histórico supuesto | Conversión 11,8 %; incidencias 1,8 % |
| MDE de conversión | +1,5 pp para planificar potencia de 80 % frente a cero |
| Umbral empresarial | Mejora de conversión superior a +1,0 pp |
| Horizonte | Reclutamiento planificado de 14 días y 7 días de maduración |
| Muestra | Se calcula antes de simular las respuestas |

### 2.3 Hipótesis y reglas previas
Conversión: H0: pB−pA ≤ 0; H1: pB−pA > 0. Alfa = 0,05 unilateral.

Relevancia empresarial: exigir límite inferior unilateral del 95 % mayor que 0,010. Esto evalúa un umbral distinto de cero y no se deduce solo de la estimación puntual.

Guardrail: H0: qB−qA ≥ 0,005; H1: qB−qA < 0,005. Exigir límite superior unilateral del 95 % menor que 0,005. No encontrar una diferencia significativa no demuestra que el guardrail esté protegido.

La recomendación combina ambos requisitos; se trata de una regla conjunta, no de escoger la prueba favorable. Los intervalos individuales no deben describirse como una región de confianza simultánea.

**Comprobación:** ¿por qué se comparan proporciones? ¿Cómo cambiarían las hipótesis si interesara cualquier diferencia, positiva o negativa?
**Respuesta:** ____________________________________.


###  Respuestas — Actividad 2

**2.1 Interpretación** (completar antes de ejecutar)

> Determinar si **el checkout simplificado de dos pasos (B) aumenta la proporción de usuarios que compran dentro de los siete días, en más de +1,0 pp respecto del checkout actual de cuatro pasos (A), sin que la incidencia de pago empeore en +0,5 pp o más**.

**Comprobación — ¿por qué se comparan proporciones? ¿Cómo cambiarían las hipótesis si interesara cualquier diferencia?**

- **Se comparan proporciones** porque la respuesta de cada usuario es **binaria** (compra / no compra) y la métrica primaria es el porcentaje de usuarios que compran. Hay solo dos niveles del factor (A y B), así que la herramienta adecuada es la diferencia de dos proporciones, no ANOVA (que se usa para respuestas cuantitativas o más de dos grupos).
- **Si interesara cualquier diferencia** (positiva o negativa), el contraste sería **bilateral**:
  - H0: pB − pA = 0
  - H1: pB − pA ≠ 0
  - El alfa de 0,05 se reparte en dos colas (0,025 por cola; z ≈ 1,96) en lugar de concentrarse en una sola (z ≈ 1,645), por lo que se necesitaría más muestra para la misma potencia.

## Actividad 3 — Preparación y planificación previa
### Paso 1. Preparar Google Colab
Ejecute las celdas en orden. La configuración y el cálculo de muestra aparecen antes de generar las respuestas. En Colab, la primera celda instala solo los paquetes que faltan. Si se solicita reiniciar, vuelva a ejecutar desde aquí.


In [ ]:
import importlib.util
import subprocess
import sys

paquetes = ["numpy", "pandas", "scipy", "statsmodels", "plotly"]
faltan = [p for p in paquetes if importlib.util.find_spec(p) is None]
if faltan:
    subprocess.check_call([sys.executable, "-m", "pip", "install", "-q", *faltan])
print("Dependencias disponibles.")


Dependencias disponibles.


In [ ]:
import numpy as np
import pandas as pd
import plotly.express as px
from scipy.stats import norm
from statsmodels.stats.power import NormalIndPower
from statsmodels.stats.proportion import proportion_effectsize

pd.set_option("display.max_columns", 20)
print("Entorno listo.")


Entorno listo.


### Paso 2. Fijar los parámetros antes de observar
Los baselines son supuestos históricos del caso ficticio, no estimaciones obtenidas de los grupos que analizaremos. El margen del guardrail se fija por criterio de negocio. Se calcula muestra para conversión y, por aproximación normal, para no inferioridad del guardrail suponiendo diferencia verdadera cero. Se toma el mayor tamaño. Las potencias se refieren a cada objetivo de diseño; no garantizan 80 % para la regla conjunta ni para superar el umbral empresarial.


In [ ]:
ALPHA = 0.05
POTENCIA = 0.80
BASELINE = 0.118
MDE = 0.015
UMBRAL_NEGOCIO = 0.010
BASE_GUARDRAIL = 0.018
MARGEN_GUARDRAIL = 0.005
SEMILLA = 20260921
DIAS_RECLUTAMIENTO = 14
VENTANA_DIAS = 7

assert 0 < BASELINE < BASELINE + MDE < 1
assert 0 < UMBRAL_NEGOCIO < MDE
analysis = NormalIndPower()
es = proportion_effectsize(BASELINE + MDE, BASELINE)
n_conversion = int(np.ceil(analysis.solve_power(
    effect_size=es, alpha=ALPHA, power=POTENCIA,
    ratio=1, alternative="larger")))
z_alpha = norm.ppf(1 - ALPHA)
z_power = norm.ppf(POTENCIA)
n_guardrail = int(np.ceil(
    2 * BASE_GUARDRAIL * (1 - BASE_GUARDRAIL)
    * (z_alpha + z_power)**2 / MARGEN_GUARDRAIL**2))
N_POR_GRUPO = max(n_conversion, n_guardrail)
print("Muestra por grupo para conversión:", n_conversion)
print("Muestra aproximada por grupo para guardrail:", n_guardrail)
print("Muestra planificada por grupo:", N_POR_GRUPO)
print("Analizar una sola vez al completar muestra y seguimiento.")


Muestra por grupo para conversión: 6027
Muestra aproximada por grupo para guardrail: 8743
Muestra planificada por grupo: 8743
Analizar una sola vez al completar muestra y seguimiento.


### Pausa de análisis 1
1. ¿Por qué el baseline debe establecerse antes de ver el resultado?
2. ¿Por qué la muestra final considera también el guardrail?
3. ¿La potencia calculada garantiza demostrar una mejora superior a 1,0 pp?
**Respuestas:** ____________________________________.


###  Respuestas — Pausa de análisis 1

1. **¿Por qué el baseline debe establecerse antes de ver el resultado?** Porque es un supuesto de planificación (proviene del histórico) que se usa para calcular la muestra y la potencia. Si se tomara de los propios grupos del experimento, el diseño quedaría calibrado con el resultado (razonamiento circular) y se podría ajustar hasta obtener la conclusión deseada. Fijarlo antes mantiene el plan independiente de los datos.

2. **¿Por qué la muestra final considera también el guardrail?** Porque la decisión exige *dos* cosas: mejorar la conversión y proteger el guardrail. Con 6 027 usuarios por grupo bastaba para la conversión, pero el guardrail (margen de solo 0,5 pp sobre una incidencia base de 1,8 %) necesita **8 743 por grupo**. Si se dimensionara solo para la conversión, la evaluación del guardrail tendría potencia insuficiente. Por eso se toma el mayor de los dos tamaños: `max(6027, 8743) = 8743`.

3. **¿La potencia calculada garantiza demostrar una mejora superior a 1,0 pp?** **No.** La potencia de 80 % se calculó para detectar el MDE (+1,5 pp) *frente a cero*, no para demostrar que el efecto supera el umbral empresarial de +1,0 pp. Para demostrar con evidencia suficiente que el efecto supera 1,0 pp (límite inferior > 1,0 pp) se requeriría una muestra mucho mayor: aproximadamente 54 000 usuarios por grupo si el efecto verdadero fuera +1,5 pp (cálculo aproximado con varianza no combinada), frente a los 8 743 planificados.

### Paso 3. Generar y validar el conjunto académico
El simulador asigna aleatoriamente el mismo número de usuarios a A y B. Genera una fila por usuario y completa siete días de observación para todos. Supone conversiones de 11,8 % y 13,8 % e incidencias de 1,8 % en ambos grupos. Son parámetros del generador, no resultados que deban recuperarse exactamente. Las respuestas se simulan independientes para simplificar el ejercicio; una empresa real puede presentar dependencias y estacionalidad no representadas aquí.


In [ ]:
rng = np.random.default_rng(SEMILLA)
total = 2 * N_POR_GRUPO
grupos = rng.permutation(np.repeat(["A_Control", "B_Tratamiento"], N_POR_GRUPO))
prob_compra = np.where(grupos == "A_Control", 0.118, 0.138)
df = pd.DataFrame({
    "usuario_id": np.arange(1, total + 1),
    "grupo": grupos,
    "dia_asignacion": rng.integers(1, DIAS_RECLUTAMIENTO + 1, total),
    "dias_observados": np.full(total, VENTANA_DIAS),
    "compra_7d": rng.binomial(1, prob_compra),
    "incidencia_pago_7d": rng.binomial(1, BASE_GUARDRAIL, total)
})
print("DATOS SIMULADOS — NexaCommerce")
print("Dimensión:", df.shape)
display(df.head())


DATOS SIMULADOS — NexaCommerce
Dimensión: (17486, 6)


,usuario_id,grupo,dia_asignacion,dias_observados,compra_7d,incidencia_pago_7d
0,1,A_Control,8,7,0,0
1,2,B_Tratamiento,5,7,0,0
2,3,A_Control,12,7,0,0
3,4,A_Control,9,7,0,0
4,5,B_Tratamiento,12,7,0,0


In [ ]:
requeridas = {"usuario_id", "grupo", "dias_observados",
              "compra_7d", "incidencia_pago_7d"}
assert requeridas.issubset(df.columns), "Faltan columnas"
assert not df[list(requeridas)].isna().any().any(), "Hay datos ausentes"
assert df["usuario_id"].is_unique, "Usuarios repetidos o contaminados"
assert set(df["grupo"]) == {"A_Control", "B_Tratamiento"}
assert (df["dias_observados"] == VENTANA_DIAS).all()
for col in ["compra_7d", "incidencia_pago_7d"]:
    assert df[col].isin([0, 1]).all(), f"{col} no es binaria"
ab = df.copy()
tam_grupos = ab.groupby("grupo").size()
assert (tam_grupos == N_POR_GRUPO).all()
display(tam_grupos.to_frame("usuarios"))
print("Calidad estructural y seguimiento completos.")


,usuarios
grupo,
A_Control,8743
B_Tratamiento,8743


Calidad estructural y seguimiento completos.


### Pausa de análisis 2
1. Registre el tamaño de cada grupo. ¿Cumple el plan?
2. ¿Qué ocurriría si un usuario apareciera en ambas variantes?
3. ¿Por qué no debe excluirse a quienes no compraron?
4. ¿Por qué es necesario completar los siete días antes de analizar?
**Respuestas:** ____________________________________.


###  Respuestas — Pausa de análisis 2

1. **Tamaño de cada grupo:** A_Control = **8 743** usuarios y B_Tratamiento = **8 743** usuarios (total 17 486). **Sí cumple el plan** (`N_POR_GRUPO = 8743`), y la validación estructural pasó: sin datos ausentes, `usuario_id` único, respuestas binarias y 7 días observados para todos.

2. **¿Qué ocurriría si un usuario apareciera en ambas variantes?** Habría **contaminación de grupos**: la respuesta de ese usuario estaría influida por ambas versiones, no se podría atribuir su compra a A o a B y se violaría la independencia entre unidades. Además, el efecto tendería a diluirse (sesgo hacia cero) y el error estándar calculado sería incorrecto. Cada usuario debe recibir únicamente la variante asignada.

3. **¿Por qué no debe excluirse a quienes no compraron?** Porque son parte del denominador de la métrica (usuarios que compraron / **todos** los asignados). Excluirlos equivaldría a condicionar por el resultado: la conversión sería 100 % en ambos grupos o quedaría sesgada, y no habría nada que comparar. Los no compradores contienen justamente la información de cuántos usuarios *no* convirtieron con cada variante.

4. **¿Por qué es necesario completar los siete días antes de analizar?** Porque la métrica se define como compra **dentro de los siete días** posteriores a la asignación. Un usuario asignado el día 14 aún no ha tenido esa ventana completa; si se analizara antes, se subestimaría la conversión y se mezclarían usuarios con distinta exposición temporal (además de incurrir en *peeking*). Con ventana completa para todos, la comparación es justa y comparable.

## Actividad 4 — MDE, potencia y sensibilidad de muestra
### Paso 4. Recuperar el diseño fijado
Revise los parámetros sin recalibrarlos con el resultado observado. El MDE sirve para dimensionar la muestra; el umbral empresarial se utiliza para valorar la mejora. En este caso, las dos cantidades difieren deliberadamente.


In [ ]:
display(pd.DataFrame({
    "parametro": ["Baseline", "MDE", "Umbral empresarial",
                  "Margen guardrail", "Alfa", "Potencia"],
    "valor": [BASELINE, MDE, UMBRAL_NEGOCIO,
              MARGEN_GUARDRAIL, ALPHA, POTENCIA]
}))
print("Muestra final por grupo:", N_POR_GRUPO)


,parametro,valor
0,Baseline,0.118
1,MDE,0.015
2,Umbral empresarial,0.010
3,Margen guardrail,0.005
4,Alfa,0.050
5,Potencia,0.800


Muestra final por grupo: 8743


### Paso 5. Explorar la sensibilidad al MDE
Compare MDE de 0,5, 1,5 y 2,5 pp. Esta exploración es didáctica y no cambia el plan ni la decisión del caso ya fijado.
**Predicción:** ¿qué efecto necesitará más muestra y por qué?


###  Respuesta — Predicción (Paso 5)

**¿Qué efecto necesitará más muestra y por qué?** El **MDE de 0,5 pp** (el más pequeño). Detectar un efecto pequeño es más difícil porque queda "enterrado" en la variabilidad aleatoria del muestreo; para distinguirlo se necesita un error estándar más pequeño, y eso solo se logra con más datos. El tamaño de muestra crece aproximadamente con el inverso del cuadrado del efecto (n ∝ 1/Δ²). Se verifica con la tabla de la celda siguiente.

In [ ]:
filas = []
for mde in [0.005, 0.015, 0.025]:
    es = proportion_effectsize(BASELINE + mde, BASELINE)
    n = analysis.solve_power(effect_size=es, alpha=ALPHA,
                             power=POTENCIA, alternative="larger")
    filas.append({"MDE_pp": mde * 100,
                  "n_conversion": int(np.ceil(n)),
                  "n_con_guardrail": max(int(np.ceil(n)), n_guardrail)})
sensibilidad_n = pd.DataFrame(filas)
display(sensibilidad_n)
fig = px.line(sensibilidad_n, x="MDE_pp",
              y=["n_conversion", "n_con_guardrail"], markers=True,
              title="Muestra por grupo según el MDE",
              labels={"value": "Usuarios por grupo", "MDE_pp": "MDE (pp)"})
fig.show()


,MDE_pp,n_conversion,n_con_guardrail
0,0.5,52413,52413
1,1.5,6027,8743
2,2.5,2241,8743


### Pausa de análisis 3
1. Explique por qué reducir el MDE aumenta la muestra para conversión.
2. ¿Por qué el guardrail puede determinar la muestra final?
3. ¿Sería válido aumentar el MDE después de ver los resultados para declarar éxito?
**Respuestas:** ____________________________________.


###  Respuestas — Pausa de análisis 3

Tabla obtenida en la sensibilidad:

| MDE (pp) | n conversión | n con guardrail |
|---:|---:|---:|
| 0,5 | 52 413 | 52 413 |
| 1,5 | 6 027 | 8 743 |
| 2,5 | 2 241 | 8 743 |

1. **¿Por qué reducir el MDE aumenta la muestra para conversión?** Porque un efecto menor es más difícil de separar del ruido aleatorio. Como n ∝ 1/Δ², al pasar de 1,5 pp a 0,5 pp (tres veces menor) la muestra crece cerca de nueve veces (6 027 → 52 413, ≈ 8,7×); y de 2,5 pp a 0,5 pp (cinco veces menor) crece ≈ 23 veces (2 241 → 52 413).

2. **¿Por qué el guardrail puede determinar la muestra final?** Porque el guardrail tiene su propio requerimiento (8 743 por grupo) que **no depende del MDE de conversión**: depende de su margen (0,5 pp) y de su incidencia base (1,8 %). Cuando el MDE de conversión es grande (1,5 o 2,5 pp), la conversión necesita menos usuarios (6 027 o 2 241) que el guardrail, y la muestra final es `max(...)` = 8 743. Solo con un MDE muy pequeño (0,5 pp) la conversión pasa a ser el requisito dominante.

3. **¿Sería válido aumentar el MDE después de ver los resultados para declarar éxito?** **No.** El MDE se fija antes de recoger datos para dimensionar la muestra; cambiarlo a posteriori para que el resultado "encaje" es manipular la regla de decisión (*HARKing*), y además no altera los datos ni la evidencia: solo altera cómo se cuenta la historia. Un MDE mayor tampoco demostraría nada sobre el efecto real.

## Actividad 5 — Análisis estadístico de la comparación A/B

**Propósito:** calcular el efecto observado y evaluar la hipótesis mediante una prueba para dos proporciones.

### Paso 6. Obtener tasas y tamaño del efecto observado

Antes de ejecutar, recuerde:

- Diferencia absoluta: $p_B-p_A$.
- Lift relativo: $(p_B-p_A)/p_A$.
- La diferencia se expresa en puntos porcentuales; el lift expresa un cambio relativo respecto de A.


In [ ]:
tabla = ab.groupby("grupo")["compra_7d"].agg(["sum", "count", "mean"])
tabla.columns = ["compras", "n", "tasa_conversion"]
x_A, n_A = int(tabla.loc["A_Control", "compras"]), int(tabla.loc["A_Control", "n"])
x_B, n_B = int(tabla.loc["B_Tratamiento", "compras"]), int(tabla.loc["B_Tratamiento", "n"])
p_A, p_B = x_A / n_A, x_B / n_B
efecto = p_B - p_A
lift_rel = efecto / p_A if p_A > 0 else np.nan
display(tabla)
print(f"Diferencia B−A: {efecto*100:.3f} pp")
print(f"Lift relativo: {lift_rel:.2%}")


,compras,n,tasa_conversion
grupo,,,
A_Control,996,8743,0.11392
B_Tratamiento,1236,8743,0.14137


Diferencia B−A: 2.745 pp
Lift relativo: 24.10%


### Pausa de análisis 4

Complete sin utilizar todavía el p-valor:

> La tasa de B es ______ y la tasa de A es ______. La diferencia observada es ______ puntos porcentuales, equivalente a un lift relativo de ______. Estos valores describen la muestra, pero todavía no permiten afirmar que ____________________________________________.


###  Respuesta — Pausa de análisis 4

> La tasa de B es **14,14 %** (1 236 compras de 8 743) y la tasa de A es **11,39 %** (996 compras de 8 743). La diferencia observada es **2,745** puntos porcentuales, equivalente a un lift relativo de **24,10 %**. Estos valores describen la muestra, pero todavía no permiten afirmar que **la diferencia no se deba al azar del muestreo, es decir, que B realmente mejora la conversión en la población de usuarios (ni que esa mejora supere el umbral empresarial de 1,0 pp)**; para ello se necesita el contraste de hipótesis y los intervalos.

### Paso 7. Contraste e intervalos coherentes
Usamos la aproximación normal de Wald con error estándar no combinado. Se comprueba que haya al menos diez compras y diez no compras en cada grupo; esta regla práctica no sustituye la revisión del diseño.

Para la prueba unilateral frente a cero, p < 0,05 equivale a límite inferior unilateral del 95 % > 0 con el mismo método. El IC bilateral del 95 % se presenta aparte. La función también permite contrastar un umbral distinto de cero. El p-valor no es la probabilidad de que H0 sea verdadera.


In [ ]:
def diferencia_wald(x_b, n_b, x_a, n_a, umbral=0.0):
    if min(x_b, n_b-x_b, x_a, n_a-x_a) < 10:
        raise ValueError("Frecuencias pequeñas: revisar método antes de continuar")
    pb, pa = x_b/n_b, x_a/n_a
    d = pb-pa
    se = np.sqrt(pb*(1-pb)/n_b + pa*(1-pa)/n_a)
    z = (d-umbral)/se
    return {"d": d, "se": se, "z": z,
            "p_mayor": norm.sf(z), "p_menor": norm.cdf(z),
            "li_uni95": d-norm.ppf(1-ALPHA)*se,
            "ls_uni95": d+norm.ppf(1-ALPHA)*se,
            "ic95_low": d-norm.ppf(1-ALPHA/2)*se,
            "ic95_high": d+norm.ppf(1-ALPHA/2)*se}

conv = diferencia_wald(x_B, n_B, x_A, n_A)
negocio = diferencia_wald(x_B, n_B, x_A, n_A, UMBRAL_NEGOCIO)
print(f"z frente a cero: {conv['z']:.4f}")
print(f"p unilateral frente a cero: {conv['p_mayor']:.6g}")
print(f"Límite inferior unilateral 95%: {conv['li_uni95']*100:.3f} pp")
print(f"IC bilateral 95%: [{conv['ic95_low']*100:.3f}, {conv['ic95_high']*100:.3f}] pp")
print(f"p unilateral frente a +1,0 pp: {negocio['p_mayor']:.6g}")


z frente a cero: 5.4436
p unilateral frente a cero: 2.61091e-08
Límite inferior unilateral 95%: 1.916 pp
IC bilateral 95%: [1.757, 3.733] pp
p unilateral frente a +1,0 pp: 0.000269553


In [ ]:
grafico = tabla.reset_index()
grafico["conversion_pct"] = grafico["tasa_conversion"] * 100
fig = px.bar(grafico, x="grupo", y="conversion_pct", text="conversion_pct",
             title="Conversión observada — datos simulados",
             labels={"conversion_pct": "Usuarios que compraron (%)", "grupo": "Grupo"})
fig.update_traces(texttemplate="%{text:.2f}%", textposition="outside")
fig.update_yaxes(rangemode="tozero")
fig.show()


### Análisis y conclusión estadística
1. Registre alfa, z y p unilateral frente a cero.
2. ¿Se rechaza H0? Explique en el contexto de conversión del caso simulado.
3. ¿El límite inferior unilateral supera cero? ¿Y el umbral de 1,0 pp?
4. Explique por qué el IC bilateral del 95 % y el límite unilateral del 95 % tienen extremos distintos.
5. ¿Superar 1,0 pp con la estimación puntual demuestra superar ese umbral con evidencia suficiente?
**Respuestas:** ____________________________________.


###  Respuestas — Análisis y conclusión estadística

1. **Alfa, z y p unilateral frente a cero:** α = **0,05** (unilateral); z = **5,4436**; p unilateral = **2,61 × 10⁻⁸** (≈ 0,0000000261).

2. **¿Se rechaza H0?** **Sí.** Como p (2,61 × 10⁻⁸) < α (0,05), se rechaza H0: pB − pA ≤ 0. En el contexto del caso simulado, hay evidencia estadística de que la conversión a siete días con el checkout simplificado (B) es mayor que con el checkout actual (A). El p-valor **no** es la probabilidad de que H0 sea verdadera: es la probabilidad de observar una diferencia tan grande o mayor si H0 fuera cierta.

3. **¿El límite inferior unilateral supera cero? ¿Y el umbral de 1,0 pp?** El límite inferior unilateral del 95 % es **1,916 pp**. Supera cero (1,916 > 0) y también supera el umbral empresarial de **1,0 pp** (1,916 > 1,0). La prueba directa frente a +1,0 pp lo confirma: p = **0,00027** < 0,05.

4. **¿Por qué el IC bilateral del 95 % y el límite unilateral del 95 % tienen extremos distintos?** Usan el mismo error estándar (≈ 0,504 pp), pero distinto valor crítico. El IC bilateral reparte el 5 % de error en dos colas (2,5 % + 2,5 %; z = 1,96) y da un intervalo **[1,757; 3,733] pp**. El límite unilateral concentra todo el 5 % en una sola cola (z = 1,645) y solo acota un lado (el inferior), por lo que queda **más cerca de la estimación** (1,916 pp) que el extremo inferior bilateral (1,757 pp). Responden preguntas distintas: el bilateral describe la incertidumbre en ambos sentidos; el unilateral responde "¿cuánto es, como mínimo, la mejora?".

5. **¿Superar 1,0 pp con la estimación puntual demuestra superar ese umbral con evidencia suficiente?** **No.** La estimación puntual (2,745 pp) es solo el resultado de una muestra y ya incluye ruido de muestreo. Superar el umbral con evidencia suficiente exige que el **límite inferior** del intervalo (o el contraste frente al umbral) lo supere. Aquí sí se cumple (1,916 > 1,0), pero por el límite y no por la estimación puntual: con una estimación de, por ejemplo, 1,2 pp y un error estándar grande, el límite inferior podría quedar por debajo de 1,0 pp.

**Conclusión estadística:** con datos simulados, B mejora la conversión de forma estadísticamente significativa (+2,745 pp; lift 24,10 %) y con evidencia suficiente de superar el umbral de +1,0 pp (límite inferior 1,916 pp). Estos resultados aún no bastan para decidir: falta evaluar el guardrail.

## Actividad 6 — Del resultado estadístico a la decisión
### Paso 8. Evaluar el guardrail con incertidumbre
El guardrail es la proporción de usuarios con alguna incidencia de pago en siete días. El margen de deterioro admisible es +0,5 pp. La no inferioridad exige que el límite superior unilateral del 95 % esté por debajo de ese margen. No basta obtener p > 0,05 en una prueba de igualdad.


In [ ]:
guardrail = ab.groupby("grupo")["incidencia_pago_7d"].agg(["sum", "count", "mean"])
guardrail.columns = ["incidencias", "n", "tasa_incidencia"]
g_A = int(guardrail.loc["A_Control", "incidencias"])
g_B = int(guardrail.loc["B_Tratamiento", "incidencias"])
guard = diferencia_wald(g_B, n_B, g_A, n_A, MARGEN_GUARDRAIL)
display(guardrail)
print(f"Cambio del guardrail B−A: {guard['d']*100:.3f} pp")
print(f"Límite superior unilateral 95%: {guard['ls_uni95']*100:.3f} pp")
print(f"Margen permitido: {MARGEN_GUARDRAIL*100:.1f} pp")
print(f"p de no inferioridad: {guard['p_menor']:.6g}")


,incidencias,n,tasa_incidencia
grupo,,,
A_Control,139,8743,0.015898
B_Tratamiento,176,8743,0.020130


Cambio del guardrail B−A: 0.423 pp
Límite superior unilateral 95%: 0.754 pp
Margen permitido: 0.5 pp
p de no inferioridad: 0.351286


###  Interpretación — Guardrail (Paso 8)

- Incidencia de pago: A = **1,59 %** (139 de 8 743); B = **2,01 %** (176 de 8 743). Cambio B−A = **+0,423 pp**.
- Límite superior unilateral del 95 % = **0,754 pp**, que **no es inferior** al margen permitido de **0,5 pp** (0,754 > 0,5).
- p de no inferioridad = **0,351** (muy por encima de 0,05): no se puede rechazar H0: qB − qA ≥ 0,5 pp.

**Lectura:** el guardrail **no queda demostrado como protegido**. El aumento estimado (+0,423 pp) es menor que el margen, pero la incertidumbre es tal que valores por encima de +0,5 pp siguen siendo compatibles con los datos. Esto es un resultado **no concluyente**: no demuestra que B perjudique los pagos ni que los proteja.

### Paso 9. Integrar la regla de decisión
| Criterio | Evidencia requerida |
|---|---|
| Diseño y calidad | Una fila por usuario, asignación persistente, muestra y seguimiento completos |
| Mejora frente a cero | p unilateral < 0,05 |
| Relevancia empresarial | Límite inferior unilateral del 95 % > +1,0 pp |
| Guardrail | Límite superior unilateral del 95 % < +0,5 pp |

La regla se fijó antes de simular. El MDE dimensiona la muestra y no reemplaza al umbral empresarial. Si no se satisface un requisito, distinguir evidencia insuficiente de evidencia de perjuicio. Cualquier experimento posterior necesita un nuevo plan; no seguir agregando datos hasta lograr significancia.


In [ ]:
cumple_estadistica = bool(conv["p_mayor"] < ALPHA)
cumple_negocio = bool(conv["li_uni95"] > UMBRAL_NEGOCIO)
cumple_guardrail = bool(guard["ls_uni95"] < MARGEN_GUARDRAIL)
cumple_diseno = bool((tam_grupos >= N_POR_GRUPO).all())
criterios = pd.Series({"Diseño validado": cumple_diseno,
    "Mejora frente a cero": cumple_estadistica,
    "Mejora superior al umbral": cumple_negocio,
    "Guardrail protegido": cumple_guardrail})
display(criterios.to_frame("cumple"))
if criterios.all():
    decision = "Recomendar B en el caso simulado, con seguimiento posterior."
else:
    decision = "No recomendar el despliegue todavía: revisar los criterios no cumplidos."
print(decision)
print("Conclusión académica con datos simulados; no valida un despliegue real.")


,cumple
Diseño validado,True
Mejora frente a cero,True
Mejora superior al umbral,True
Guardrail protegido,False


No recomendar el despliegue todavía: revisar los criterios no cumplidos.
Conclusión académica con datos simulados; no valida un despliegue real.


###  Interpretación — Regla de decisión (Paso 9)

| Criterio | ¿Cumple? | Evidencia |
|---|:---:|---|
| Diseño validado | ✅ Sí | 8 743 usuarios por grupo, una fila por usuario, 7 días completos |
| Mejora frente a cero | ✅ Sí | p = 2,61 × 10⁻⁸ < 0,05 |
| Mejora superior al umbral | ✅ Sí | límite inferior 1,916 pp > 1,0 pp |
| Guardrail protegido | ❌ No | límite superior 0,754 pp ≥ 0,5 pp |

Como la regla es **conjunta** y un requisito no se cumple, el resultado es: **no recomendar el despliegue todavía**.

### Pausa de decisión
| Componente | Resultado observado | Interpretación |
|---|---|---|
| p unilateral frente a cero | | |
| Diferencia B−A en pp y lift relativo | | |
| Límite inferior y umbral empresarial | | |
| Límite superior del guardrail y margen | | |
| Calidad y límites del caso simulado | | |

**Decisión en cuatro líneas:** recomendar B / no recomendar todavía, porque ___________. Incluya al menos dos cifras. Si falta evidencia, proponga un nuevo diseño válido; no continúe esta muestra buscando significancia.


###  Respuestas — Pausa de decisión

| Componente | Resultado observado | Interpretación |
|---|---|---|
| p unilateral frente a cero | 2,61 × 10⁻⁸ (z = 5,44) | Es extremadamente improbable observar una diferencia tan grande si B no mejorara a A; se rechaza H0. No es la probabilidad de que H0 sea cierta. |
| Diferencia B−A en pp y lift relativo | +2,745 pp (B 14,14 % vs A 11,39 %); lift 24,10 % | Mejora grande en la muestra, superior al MDE (1,5 pp) y al umbral (1,0 pp). Es mayor que los +2,0 pp con que se generaron los datos: refleja variabilidad de muestreo. |
| Límite inferior y umbral empresarial | LI unilateral 95 % = 1,916 pp vs umbral 1,0 pp (p frente al umbral = 0,00027) | Hay evidencia suficiente de que la mejora supera el umbral de relevancia práctica. |
| Límite superior del guardrail y margen | Δ incidencias = +0,423 pp; LS unilateral 95 % = 0,754 pp vs margen 0,5 pp; p de no inferioridad = 0,351 | No se demuestra que el guardrail esté protegido. Resultado no concluyente: ni protección demostrada ni perjuicio demostrado. |
| Calidad y límites del caso simulado | 8 743 por grupo, sin ausentes, IDs únicos, ventana de 7 días completa | Diseño válido, pero los datos son sintéticos, con respuestas independientes y supuestos académicos: la conclusión no valida un despliegue real. |

**Decisión en cuatro líneas:**

**No recomendar B todavía**, porque aunque mejora la conversión con evidencia suficiente (+2,745 pp; límite inferior 1,916 pp > umbral de 1,0 pp),
el guardrail de incidencias de pago no queda protegido: el aumento es de +0,423 pp y su límite superior (0,754 pp) supera el margen de 0,5 pp.
La evidencia sobre el guardrail es insuficiente (no concluyente), no evidencia de perjuicio; no se debe seguir agregando datos a esta muestra hasta lograr significancia.
Propuesta: diseñar un **nuevo experimento con plan prerregistrado** (misma métrica y regla, muestra dimensionada para el guardrail con mayor precisión, por ejemplo cerca de 24 000 por grupo si se espera un aumento verdadero de +0,2 pp), y revisar en paralelo por qué el checkout simplificado podría afectar los pagos.

## Actividad 7 — Retos de aplicación y retroalimentación
Se resuelven durante la clase y no generan una entrega adicional.

### Reto 1 — Estimación y umbral
Un caso presenta p = 0,02 frente a cero, mejora estimada de 0,4 pp y umbral empresarial de 3 pp. ¿Qué demuestra la prueba y qué no demuestra? ¿Qué información aporta el intervalo?
**Respuesta:** ____________________________________.

### Reto 2 — Cambio de pregunta
Si interesa cualquier diferencia entre A y B, formule H0 y H1 e indique el tipo de contraste. ¿Por qué debe elegirse antes de observar los resultados?
**Respuesta:** ____________________________________.

### Reto 3 — Guardrail no concluyente
El margen es +0,5 pp. El aumento estimado de incidencias es +0,2 pp, pero el límite superior unilateral es +0,7 pp. ¿Se ha demostrado cumplir el guardrail? ¿Es correcto afirmar que se ha demostrado perjuicio?
**Respuesta:** ____________________________________.

### Reto 4 — Reproducibilidad y sesgo
Un equipo cambia la semilla hasta conseguir una decisión favorable. Explique por qué eso invalida el ejercicio. Proponga cómo documentar una exploración de escenarios sin sustituir el caso base.
**Respuesta:** ____________________________________.


###  Respuestas — Actividad 7 (Retos de aplicación)

**Reto 1 — Estimación y umbral**

- **Qué demuestra la prueba:** con p = 0,02 frente a cero hay evidencia estadística de que el efecto es mayor que cero (o distinto de cero, según el contraste); es decir, es improbable que la mejora sea exactamente nula.
- **Qué no demuestra:** que la mejora sea **relevante en la práctica**. La estimación (0,4 pp) es muy inferior al umbral empresarial (3 pp); la significancia estadística no equivale a relevancia práctica (con muestras grandes efectos pequeños pueden ser significativos).
- **Qué aporta el intervalo:** muestra el rango de tamaños de efecto plausibles. Con p = 0,02 y una estimación de 0,4 pp, el error estándar es del orden de 0,17–0,20 pp y el IC del 95 % quedaría aproximadamente entre 0,0 y 0,8 pp: **todo el intervalo está muy por debajo de 3 pp**, así que hay evidencia de que el efecto es real pero no alcanza el umbral. No se recomendaría implementar por criterio empresarial.

**Reto 2 — Cambio de pregunta**

- Hipótesis: **H0: pB − pA = 0** y **H1: pB − pA ≠ 0**, con un contraste **bilateral** (alfa repartido en ambas colas).
- Debe elegirse **antes** de observar los resultados porque el tipo de contraste determina la región de rechazo y el error tipo I. Si se decidiera después de ver, por ejemplo, que B > A, y se cambiara a unilateral, el p-valor se reduciría a la mitad artificialmente y se inflaría la probabilidad de falsos positivos (decidir la pregunta según los datos es manipular el análisis).

**Reto 3 — Guardrail no concluyente**

- **¿Se ha demostrado cumplir el guardrail?** **No.** Para demostrarlo, el límite superior unilateral debía ser menor que +0,5 pp y es +0,7 pp; valores sobre el margen siguen siendo compatibles con los datos.
- **¿Es correcto afirmar que se ha demostrado perjuicio?** **Tampoco.** La estimación puntual (+0,2 pp) está por debajo del margen y el intervalo incluye tanto aumentos menores como mayores que 0,5 pp. Es un resultado **no concluyente**: evidencia insuficiente no es evidencia de perjuicio, y "no significativo" no significa "igual". Faltaría más información (más muestra en un nuevo diseño) para concluir.

**Reto 4 — Reproducibilidad y sesgo**

- **Por qué invalida el ejercicio:** cambiar la semilla hasta obtener una decisión favorable es **selección de resultados** (*p-hacking* / múltiples intentos): con suficientes intentos siempre saldrá algún resultado "significativo" por azar, el error tipo I real deja de ser 5 % y la conclusión no refleja evidencia sino la búsqueda de un resultado deseado. Además, contradice la norma de no modificar ni fabricar resultados para forzar una conclusión.
- **Cómo documentar una exploración sin sustituir el caso base:** conservar la semilla del caso base (20260921) y su decisión como resultado oficial; definir **antes** el conjunto de escenarios o semillas a explorar (por ejemplo, 100 semillas o distintos supuestos de conversión); ejecutarlos **todos** y reportarlos completos (por ejemplo, proporción de veces que se cumple cada criterio); rotular la exploración como análisis de sensibilidad y registrar semilla, parámetros y código de cada corrida.

# VIII. Secuencia de trabajo para la sesión
| Momento | Minutos | Desarrollo |
|---|---:|---|
| Apertura | 10 | Pregunta orientadora y caso |
| Actividad 1 | 15 | Activación conceptual |
| Actividad 2 | 20 | Representación e hipótesis |
| Actividad 3 | 30 | Planificación, simulación y calidad |
| Receso | 15 | Pausa |
| Actividad 4 | 25 | Sensibilidad de muestra |
| Actividad 5 | 30 | Tasas, contraste e intervalos |
| Actividad 6 | 20 | Guardrail y decisión |
| Actividad 7 | 15 | Retos y ticket de salida |

Total: 180 minutos, incluido el receso.

## Ticket de salida
«Un experimento A/B no termina en el p-valor porque…».


###  Respuesta — Ticket de salida

> «Un experimento A/B no termina en el p-valor porque **el p-valor solo indica qué tan compatible es el dato con H0; no dice cuán grande es el efecto, si supera el umbral empresarial, si los guardrails están protegidos ni si el diseño fue válido. La decisión debe integrar tamaño del efecto, intervalos de confianza, relevancia práctica y guardrails**.»

En este laboratorio se ve claramente: el p-valor frente a cero fue 2,61 × 10⁻⁸ y la mejora superó el umbral (límite inferior 1,916 pp > 1,0 pp), pero aun así **no se recomienda el despliegue** porque el guardrail de incidencias no quedó protegido (límite superior 0,754 pp > 0,5 pp).

# IX. Lista de verificación de aprendizaje
- [ ] Identifico factor, niveles y unidad experimental.
- [ ] Defino respuesta, denominador y ventana de observación.
- [ ] Formulo hipótesis antes de observar resultados.
- [ ] Distingo baseline, MDE, umbral empresarial, alfa y potencia.
- [ ] Calculo muestra antes de generar los datos del caso.
- [ ] Interpreto pp, lift, p-valor e intervalos sin confundirlos.
- [ ] Evalúo el guardrail con margen e incertidumbre.
- [ ] Distingo evidencia insuficiente de evidencia de perjuicio.
- [ ] Reconozco peeking, contaminación y selección de resultados.
- [ ] Declaro el carácter simulado y los límites del ejercicio.


# X. Referencias
- Kohavi, R., Tang, D., y Xu, Y. (2020). Trustworthy Online Controlled Experiments. Cambridge University Press.
- Bruce, P., Bruce, A., y Gedeck, P. (2020). Practical Statistics for Data Scientists, segunda edición. O’Reilly.
- statsmodels. Documentación de NormalIndPower y proportion_effectsize: https://www.statsmodels.org/stable/stats.html
- SciPy. Distribución normal, funciones cdf, sf y ppf: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html

Datos: simulación académica NexaCommerce incluida en el notebook, semilla 20260921. Los datos no proceden de las referencias bibliográficas.
